### PR TITLE
Added Homer/sipcapture agent feature

### PR DESCRIPTION
**Pull request by sergey-safarov**

### CODE REVIEW

**Comment by sergey-safarov** [comment]
This requires latest kamailio
https://github.com/kamailio/kamailio/commit/f44a02672d6937deafda21f99a912fe55a3d96d6

**Comment by sergey-safarov** [comment]
Hello Luis (@lazedo) 
Please review updated PR. If ok, then i will test on this config and squash into one commit.

File: kamailio/default.cfg [hunk 1]
```diff
@@ -670,6 +674,10 @@ route[INTERNAL_TO_EXTERNAL_RELAY]
 
     remove_hf_re("^X-.*");
 
+    #!ifdef SIP_TRACE_ROLE
+    route(SEND_SIP_TRACE);
+    #!endif
+
```
Comment (lazedo) [inline kamailio/default.cfg]: @sergey-safarov can this block be moved to route[RELAY] and delete the other identical block below ?

File: kamailio/sip_trace-role.cfg [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+modparam("siptrace", "trace_to_database", 0)
+## `trace_flag` value must be equal to FLAG_SIP_TRACE value at head of "default.cfg" file
+modparam("siptrace", "trace_flag", 8)
+modparam("siptrace", "trace_on", 1)
+
+route[SEND_SIP_TRACE]
+{
+    setflag(FLAG_SIP_TRACE);
+    sip_trace();
+}
```
Comment (lazedo) [inline kamailio/sip_trace-role.cfg]: @sergey-safarov here's some additions that i would like to be added, comments appreciated. these options allow  for the role to be enabled but controls what parts it should do and can be changed at runtime  (we may also have homer capture on freeswitch).

```
kazoo.trace = KZ_TRACE descr "activates global tracing"
kazoo.trace_external = KZ_TRACE_INTERNAL descr "activates tracing from external sources"
kazoo.trace_internal = KZ_TRACE_EXTERNAL descr "activates tracing from internal sources"

kazoo.trace_internal_incoming = KZ_TRACE_INTERNAL_INCOMING descr "traces the original request as received from internal sources"
kazoo.trace_internal_outgoing = KZ_TRACE_INTERNAL_OUTGOING descr "traces the outgoing request to external sources after possible modification"
kazoo.trace_external_incoming = KZ_TRACE_EXTERNAL_INCOMING descr "traces the original request as received from external sources"
kazoo.trace_external_outgoing = KZ_TRACE_EXTERNAL_OUTGOING descr "traces the outgoing request to internal sources after possible modification"

route[SIP_TRACE_INTERNAL]
{
     if(@cfg_get.kazoo.trace_internal == 0) {
          return;
     }
    if(@cfg_get.kazoo.trace_internal_incoming  == 1) {
          sip_trace();
   }
    if(@cfg_get.kazoo.trace_internal_outgoing == 1) {
          setflag(FLAG_SIP_TRACE);
   }
}

route[SIP_TRACE_EXTERNAL]
{
.... same as above but proper selectors
}

route[SEND_SIP_TRACE]
{
     if(@cfg_get.kazoo.trace  == 0) {
          return;
     }

    if (isflagset(FLAG_INTERNALLY_SOURCED)) {
         route(SIP_TRACE_INTERNAL);
     } else {
         route(SIP_TRACE_EXTERNAL);
     }
}

```
the `KZ_TRACE` and friends should be declared in defs.cfg as #!ifndef .. #!define .. so they can be overridden in local.cfg. the default values could be something like the below (suggestions for better default welcomed)

```
#!ifndef KZ_TRACE
#!define KZ_TRACE 0
#!endif

##note : the above can probably be simplified with  a single line
## needs testing and probably apply to other integer defines in defs.cfg
`#!trydef KZ_TRACE 0`

... ommitting the preprocessor macro for simplicity
KZ_TRACE_INTERNAL 1
KZ_TRACE_EXTERNAL 1
KZ_TRACE_INTERNAL_INCOMING 1
KZ_TRACE_INTERNAL_OUTGOING 1
KZ_TRACE_EXTERNAL_INCOMING 1
KZ_TRACE_EXTERNAL_OUTGOING 1
```

**Comment by sergey-safarov** [comment]
Luis @lazedo, I can test runtime SIP_TRACE_RLE setting change. But I not know how to do it without kamailio restart.
Could you suggest how to do it.

**Comment by lazedo** [comment]
@sergey-safarov thanks. you can test with `kamcmd cfg.seti kazoo trace 1` to enable or  `kamcmd cfg.seti kazoo trace 0` to disable it. same goes with the other declared variables, `kazoo.trace_external_outgoing` will be set/unset with `kacmd cfg.seti kazoo trace_external_outgoing X`.
you can also get the current value with `kamcmd cfg.get kazoo xxxxx`

**Comment by sergey-safarov** [comment]
Hello Luis @lazedo 
I tested this version and captured signalling is not properly. Example in attached file. This is because `route[RELAY]` is called only from `route[HANDLE_IN_DIALOG_REQUESTS]`. When initial `INVITE` (external and internal) arrived to kamailio dialog is established and `route[RELAY]` is not executed.
To fix this is need to add similar calls to `route[INTERNAL_TO_EXTERNAL_RELAY]` and `route[EXTERNAL_TO_INTERNAL_RELAY]`. Next commit will fix it.
[partial-call-log.pcap.gz](https://github.com/2600hz/kazoo-configs-kamailio/files/1844469/partial-call-log.pcap.gz)

**Comment by lazedo** [comment]
@sergey-safarov thanks for the test. the problem is on main route. maybe this will work
```
...
    if (isflagset(FLAG_INTERNALLY_SOURCED)) {
        #!ifdef RESPONDER_ROLE
        route(HANDLE_RESPOND);
        #!endif
##        route(INTERNAL_TO_EXTERNAL_RELAY);
        route(RELAY);
        exit();
    }

    #!ifdef DISPATCHER_ROLE
    route(DISPATCHER_FIND_ROUTES);
    #!endif

##    route(EXTERNAL_TO_INTERNAL_RELAY);
    route(RELAY);
}
```

**Comment by sergey-safarov** [comment]
@lazedo I little rewrote your suggestions. Could you look last commit.

**Comment by lazedo** [comment]
@sergey-safarov i like your last changes. were you able to test changing trace at runtime (`kamcmd cfg.seti kazoo trace 1/0`) ?

**Comment by lazedo** [comment]
@sergey-safarov i'm ready to merge this (depending on your runtime tests), can you please  squash the commits ? thanks

**Comment by sergey-safarov** [comment]
@lazedo I tested

1. `kamcmd cfg.seti kazoo trace 1`;
2. `kamcmd cfg.seti kazoo trace_external 0`;
3. `kamcmd cfg.seti kazoo trace_internal 0`.

This works as expected.
But command `kamcmd cfg.seti kazoo trace_external_outgoing 0` and `kamcmd cfg.seti kazoo trace_external_incoming 0` and similar `internal` settings is produce miss-leading PCAP files.
I suggest to remove `trace_external_outgoing`, `trace_external_incoming`, `trace_internal_incoming ` and `trace_internal_outgoing ` functionality.
Example of pcap is attached.

[incoming-disabled.pcap.gz](https://github.com/2600hz/kazoo-configs-kamailio/files/1844540/incoming-disabled.pcap.gz)
[outgoint_disabled.pcap.gz](https://github.com/2600hz/kazoo-configs-kamailio/files/1844541/outgoint_disabled.pcap.gz)

**Comment by sergey-safarov** [comment]
Also `kamcmd cfg.seti kazoo trace 1` is equivalent of command `kamcmd siptrace.status on`.
Is possible to create alias of `siptrace.status on` to `kamcmd cfg.seti kazoo trace 1`?

**Comment by lazedo** [comment]
@sergey-safarov it may seem that it misleads but remember that these granular control is to be used with captures namely from `freeswitch`.

as for siptrace.status you can try changing
```
route[SEND_SIP_TRACE]
{
    if(@cfg_get.siptrace.status  == 0) {
        return;
    }
```
and ..
modparam("siptrace", "trace_on", KZ_TRACE)

if it works we can drop
`kazoo.trace = KZ_TRACE descr "activates global tracing"`

**Comment by sergey-safarov** [comment]
@lazedo please look 3 last commit
* Important - I found that `@cfg_get.siptrace.status` cannot be used, siptrace setting is not available via config commands. I simple removed this block.

Also added proper siptrace handling for `MESSAGE` method.

Config is tested.
If all ok for you, then i ready to squash.

File: kamailio/default.cfg [hunk 3]
```diff
@@ -621,7 +623,11 @@ route[PREPARE_INITIAL_REQUESTS]
 
 route[RELAY]
 {
-    if (isflagset(FLAG_INTERNALLY_SOURCED)) {
+    #!ifdef SIP_TRACE_ROLE
+    route(SEND_SIP_TRACE);
+    #!endif
+
+    if (isflagset(FLAG_INTERNALLY_SOURCED) || src_ip == myself) {
```
Comment (lazedo) [inline kamailio/default.cfg]: @sergey-safarov why was this added ? ` || src_ip == myself`

Comment (sergey-safarov) [reply kamailio/default.cfg]: When messages is send from kazoo_apps to kamailio, this is first send to kamailio to own socket, and then to sip device.

Comment (lazedo) [reply kamailio/default.cfg]: @sergey-safarov almost sure this will break existing stuff. if the reason was the `MESSAGE` role, then lets handle that differently in `HANDLE_MESSAGE`.

Comment (sergey-safarov) [reply kamailio/default.cfg]: Special handling for messages send kamailio to self is added.

**Comment by sergey-safarov** [comment]
@lazedo could you look last commit. Special handling for messages send kamailio to self is added.

**Comment by lazedo** [comment]
@sergey-safarov looks good, please squash the commits

**Comment by sergey-safarov** [comment]
@lazedo commits is  squashed

**Comment by lazedo** [comment]
@sergey-safarov thanks for all your work on this enhancement
